File: Tests/SplitLaneCoreTests/SOCKS5NegotiatorTests.swift
import Testing
import Foundation
@testable import SplitLaneCore

@Suite("SOCKS5 negotiator")
struct SOCKS5NegotiatorTests {

    // MARK: - Helpers

    /// Server method-selection reply.
    private func methodSelection(_ method: UInt8, version: UInt8 = 0x05) -> [UInt8] {
        [version, method]
    }

    /// Server CONNECT reply with an IPv4 bound address.
    private func connectReply(
        _ code: UInt8 = 0x00,
        version: UInt8 = 0x05,
        reserved: UInt8 = 0x00,
        address: [UInt8] = [0x01, 0, 0, 0, 0],
        port: [UInt8] = [0x00, 0x00]
    ) -> [UInt8] {
        [version, code, reserved] + address + port
    }

    private func step(_ negotiator: inout SOCKS5Negotiator, _ bytes: [UInt8]) throws -> SOCKS5Step {
        try negotiator.receive(bytes)
    }

    // MARK: - Greeting

    @Test("Greeting offers only NO AUTH when there is no credential")
    func greetingWithoutCredential() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        #expect(try negotiator.start() == .send([0x05, 0x01, 0x00]))
    }

    @Test("Greeting offers NO AUTH and username/password when a credential exists")
    func greetingWithCredential() throws {
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(username: "u", password: "p")
        )
        #expect(try negotiator.start() == .send([0x05, 0x02, 0x00, 0x02]))
    }

    @Test("start() twice is a protocol violation")
    func doubleStart() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        #expect(throws: SOCKS5Error.self) { _ = try negotiator.start() }
    }

    @Test("receive() before start() is a protocol violation")
    func receiveBeforeStart() {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        #expect(throws: SOCKS5Error.self) { _ = try negotiator.receive([0x05, 0x00]) }
    }

    // MARK: - CONNECT request encoding

    @Test("CONNECT with a domain destination is encoded per RFC 1928")
    func connectRequestDomain() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        guard case .send(let request) = try step(&negotiator, methodSelection(0x00)) else {
            Issue.record("expected a CONNECT request")
            return
        }
        let expected: [UInt8] =
            [0x05, 0x01, 0x00, 0x03, 11] + Array("example.com".utf8) + [0x01, 0xBB]
        #expect(request == expected)
    }

    @Test("CONNECT with an IPv4 destination is encoded per RFC 1928")
    func connectRequestIPv4() throws {
        var negotiator = SOCKS5Negotiator(destination: .ipv4([93, 184, 216, 34]), port: 80)
        _ = try negotiator.start()

        guard case .send(let request) = try step(&negotiator, methodSelection(0x00)) else {
            Issue.record("expected a CONNECT request")
            return
        }
        #expect(request == [0x05, 0x01, 0x00, 0x01, 93, 184, 216, 34, 0x00, 0x50])
    }

    @Test("CONNECT with an IPv6 destination is encoded per RFC 1928")
    func connectRequestIPv6() throws {
        let address = [UInt8](repeating: 0, count: 15) + [1]   // ::1
        var negotiator = SOCKS5Negotiator(destination: .ipv6(address), port: 8080)
        _ = try negotiator.start()

        guard case .send(let request) = try step(&negotiator, methodSelection(0x00)) else {
            Issue.record("expected a CONNECT request")
            return
        }
        #expect(request == [0x05, 0x01, 0x00, 0x04] + address + [0x1F, 0x90])
    }

    @Test("Ports are encoded big-endian", arguments: [
        (UInt16(443), [UInt8]([0x01, 0xBB])),
        (UInt16(80), [UInt8]([0x00, 0x50])),
        (UInt16(65535), [UInt8]([0xFF, 0xFF])),
        (UInt16(1), [UInt8]([0x00, 0x01])),
    ])
    func portEncoding(port: UInt16, expected: [UInt8]) throws {
        var negotiator = SOCKS5Negotiator(destination: .ipv4([1, 2, 3, 4]), port: port)
        _ = try negotiator.start()

        guard case .send(let request) = try step(&negotiator, methodSelection(0x00)) else {
            Issue.record("expected a CONNECT request")
            return
        }
        #expect(Array(request.suffix(2)) == expected)
    }

    // MARK: - Authentication

    @Test("Username/password sub-negotiation follows RFC 1929")
    func authenticationEncoding() throws {
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(username: "alice", password: "s3cret")
        )
        _ = try negotiator.start()

        guard case .send(let auth) = try step(&negotiator, methodSelection(0x02)) else {
            Issue.record("expected an authentication message")
            return
        }
        // Version byte is 0x01 — RFC 1929 versions itself independently of SOCKS5's 0x05.
        let expected: [UInt8] =
            [0x01, 5] + Array("alice".utf8) + [6] + Array("s3cret".utf8)
        #expect(auth == expected)
    }

    @Test("A successful auth reply is followed by the CONNECT request")
    func authenticationSuccess() throws {
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(username: "u", password: "p")
        )
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x02))

        guard case .send(let request) = try step(&negotiator, [0x01, 0x00]) else {
            Issue.record("expected a CONNECT request")
            return
        }
        #expect(request.starts(with: [0x05, 0x01, 0x00, 0x03]))
    }

    @Test("A rejected credential fails the handshake")
    func authenticationFailure() throws {
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(username: "u", password: "wrong")
        )
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x02))

        #expect(throws: SOCKS5Error.authenticationFailed) {
            _ = try negotiator.receive([0x01, 0x01])
        }
        #expect(negotiator.state == .failed)
    }

    @Test("An auth reply echoing version 0x05 is tolerated")
    func authenticationVersionTolerance() throws {
        // Some servers mistakenly echo the SOCKS version here. The status byte carries the
        // decision, so rejecting the reply outright would break interoperability for no gain.
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(username: "u", password: "p")
        )
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x02))

        guard case .send = try step(&negotiator, [0x05, 0x00]) else {
            Issue.record("expected a CONNECT request")
            return
        }
    }

    @Test("Selecting username/password without a credential fails")
    func authenticationRequiredButAbsent() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        #expect(throws: SOCKS5Error.authenticationRequired) {
            _ = try negotiator.receive(methodSelection(0x02))
        }
    }

    @Test("An over-long credential is rejected before anything is sent")
    func credentialTooLong() {
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(
                username: String(repeating: "u", count: 256),
                password: "p"
            )
        )
        #expect(throws: SOCKS5Error.credentialTooLong) { _ = try negotiator.start() }
    }

    @Test("An empty username is rejected — RFC 1929 requires at least one byte")
    func emptyUsernameRejected() {
        var negotiator = SOCKS5Negotiator(
            destination: .domain("example.com"),
            port: 443,
            credential: SOCKS5Credential(username: "", password: "p")
        )
        #expect(throws: SOCKS5Error.credentialTooLong) { _ = try negotiator.start() }
    }

    // MARK: - Method selection failures

    @Test("A server accepting no methods fails the handshake")
    func noAcceptableMethods() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        #expect(throws: SOCKS5Error.noAcceptableAuthenticationMethod) {
            _ = try negotiator.receive(methodSelection(0xFF))
        }
        #expect(negotiator.state == .failed)
    }

    @Test("A wrong version byte fails the handshake")
    func wrongVersion() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        #expect(throws: SOCKS5Error.unexpectedVersion(0x04)) {
            _ = try negotiator.receive(methodSelection(0x00, version: 0x04))
        }
    }

    @Test("An unknown method byte fails the handshake")
    func unknownMethod() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        #expect(throws: SOCKS5Error.unsupportedAuthenticationMethod(0x7F)) {
            _ = try negotiator.receive(methodSelection(0x7F))
        }
    }

    @Test("GSSAPI is refused — it is never offered")
    func gssapiRefused() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        #expect(throws: SOCKS5Error.unsupportedAuthenticationMethod(0x01)) {
            _ = try negotiator.receive(methodSelection(0x01))
        }
    }

    // MARK: - CONNECT reply

    @Test("A successful reply establishes the tunnel")
    func connectSuccess() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        guard case .established(let info) = try step(&negotiator, connectReply()) else {
            Issue.record("expected the handshake to complete")
            return
        }
        #expect(info.method == .noAuthentication)
        #expect(info.boundAddress == .ipv4([0, 0, 0, 0]))
        #expect(info.boundPort == 0)
        #expect(info.leftoverBytes.isEmpty)
        #expect(negotiator.state == .established)
    }

    @Test("Every rejection code surfaces as requestRejected", arguments: [
        (UInt8(0x01), SOCKS5.ReplyCode.generalFailure),
        (UInt8(0x02), .connectionNotAllowed),
        (UInt8(0x03), .networkUnreachable),
        (UInt8(0x04), .hostUnreachable),
        (UInt8(0x05), .connectionRefused),
        (UInt8(0x06), .ttlExpired),
        (UInt8(0x07), .commandNotSupported),
        (UInt8(0x08), .addressTypeNotSupported),
    ])
    func connectRejections(raw: UInt8, expected: SOCKS5.ReplyCode) throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        #expect(throws: SOCKS5Error.requestRejected(expected)) {
            _ = try negotiator.receive(connectReply(raw))
        }
        #expect(negotiator.state == .failed)
    }

    @Test("An out-of-range reply code is reported rather than guessed at")
    func unknownReplyCode() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        #expect(throws: SOCKS5Error.unknownReplyCode(0x42)) {
            _ = try negotiator.receive(connectReply(0x42))
        }
    }

    @Test("A non-zero reserved byte is a malformed reply")
    func nonZeroReservedByte() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        #expect(throws: SOCKS5Error.self) {
            _ = try negotiator.receive(connectReply(0x00, reserved: 0x01))
        }
    }

    @Test("A reply with a domain bound address parses")
    func replyWithDomainAddress() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        let address: [UInt8] = [0x03, 5] + Array("proxy".utf8)
        guard case .established(let info) = try step(
            &negotiator, connectReply(0x00, address: address, port: [0x1F, 0x90])
        ) else {
            Issue.record("expected the handshake to complete")
            return
        }
        #expect(info.boundAddress == .domain("proxy"))
        #expect(info.boundPort == 8080)
    }

    @Test("An unknown address type in a reply is reported")
    func unknownAddressType() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        #expect(throws: SOCKS5Error.unsupportedAddressType(0x09)) {
            _ = try negotiator.receive([0x05, 0x00, 0x00, 0x09, 0, 0])
        }
    }

    // MARK: - Streaming and truncation (F-7)

    @Test("A handshake split one byte at a time still completes")
    func byteAtATimeDelivery() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        // TCP is a stream: a two-byte reply can genuinely arrive as two reads.
        #expect(try negotiator.receive([0x05]) == .needMoreBytes)
        guard case .send = try negotiator.receive([0x00]) else {
            Issue.record("expected a CONNECT request once the reply completed")
            return
        }

        let reply = connectReply()
        for byte in reply.dropLast() {
            #expect(try negotiator.receive([byte]) == .needMoreBytes)
        }
        guard case .established = try negotiator.receive([reply[reply.count - 1]]) else {
            Issue.record("expected the handshake to complete on the final byte")
            return
        }
    }

    @Test("Truncation at every offset yields needMoreBytes, never a crash")
    func truncationAtEveryOffset() throws {
        let reply = connectReply(0x00, address: [0x03, 11] + Array("example.com".utf8))

        // The whole point of ByteReader: a hostile or broken proxy can truncate anywhere, and
        // every prefix must be a clean "wait", not a trap. A trap here kills the provider, which
        // fails every selected app open.
        for prefixLength in 0..<reply.count {
            var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
            _ = try negotiator.start()
            _ = try negotiator.receive(methodSelection(0x00))

            let step = try negotiator.receive(Array(reply.prefix(prefixLength)))
            #expect(step == .needMoreBytes, "prefix of \(prefixLength) bytes should need more")
        }
    }

    @Test("An oversized domain length in a reply cannot over-read")
    func oversizedDomainLength() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        // Claims 255 bytes of domain, supplies 3. Must wait, not read past the buffer.
        let malicious: [UInt8] = [0x05, 0x00, 0x00, 0x03, 255, 0x41, 0x41, 0x41]
        #expect(try negotiator.receive(malicious) == .needMoreBytes)
    }

    @Test("A zero-length domain in a reply is handled")
    func zeroLengthDomain() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        guard case .established(let info) = try negotiator.receive(
            [0x05, 0x00, 0x00, 0x03, 0x00, 0x00, 0x50]
        ) else {
            Issue.record("expected the handshake to complete")
            return
        }
        #expect(info.boundAddress == .domain(""))
    }

    @Test("Invalid UTF-8 in a domain reply is a malformed response, not a crash")
    func invalidUTF8Domain() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        #expect(throws: SOCKS5Error.self) {
            _ = try negotiator.receive([0x05, 0x00, 0x00, 0x03, 0x02, 0xFF, 0xFE, 0x00, 0x50])
        }
    }

    @Test("Bytes coalesced after the reply are returned, not dropped")
    func leftoverBytesArePreserved() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))

        // A server may pack its reply and the first application bytes into one segment. Dropping
        // the tail would look like a rare, unreproducible truncation of the first response.
        let payload: [UInt8] = Array("HTTP/1.1 200 OK".utf8)
        guard case .established(let info) = try negotiator.receive(connectReply() + payload) else {
            Issue.record("expected the handshake to complete")
            return
        }
        #expect(info.leftoverBytes == payload)
    }

    @Test("An entire handshake delivered in one buffer works")
    func fullyCoalescedHandshake() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()

        // The method selection and CONNECT reply cannot legitimately be coalesced, since the
        // client must send the request in between — but a buffered method selection followed
        // immediately by a reply exercises the buffer-retention path.
        guard case .send = try negotiator.receive(methodSelection(0x00)) else {
            Issue.record("expected a CONNECT request")
            return
        }
        guard case .established = try negotiator.receive(connectReply()) else {
            Issue.record("expected the handshake to complete")
            return
        }
    }

    @Test("receive() after the handshake completed is a protocol violation")
    func receiveAfterEstablished() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try step(&negotiator, methodSelection(0x00))
        _ = try step(&negotiator, connectReply())

        #expect(throws: SOCKS5Error.self) { _ = try negotiator.receive([0x00]) }
    }

    @Test("receive() after a failure is a protocol violation")
    func receiveAfterFailure() throws {
        var negotiator = SOCKS5Negotiator(destination: .domain("example.com"), port: 443)
        _ = try negotiator.start()
        _ = try? negotiator.receive(methodSelection(0xFF))

        #expect(negotiator.state == .failed)
        #expect(throws: SOCKS5Error.self) { _ = try negotiator.receive([0x05, 0x00]) }
    }
}

File: windows/tests/SplitLane.Core.Tests/Socks5NegotiatorTests.cs
using System.Text;
using SplitLane.Core.Proxy.Socks5;

namespace SplitLane.Core.Tests;

/// <summary>
/// The SOCKS5 handshake, driven byte by byte.
/// </summary>
/// <remarks>
/// Because the negotiator is a pure state machine, every truncation offset and every malformed field
/// is reachable from a unit test with no socket. That is the entire reason it is shaped this way.
/// </remarks>
public sealed class Socks5NegotiatorTests
{
    private static Socks5Negotiator ForExample(Socks5Credential? credential = null)
        => new(Socks5Address.FromDomain("example.com"), 443, credential);

    private static byte[] MethodSelection(Socks5Method method) => [Socks5.Version, (byte)method];

    private static byte[] ConnectReply(Socks5ReplyCode code = Socks5ReplyCode.Succeeded)
        => [Socks5.Version, (byte)code, 0x00, (byte)Socks5AddressType.IPv4, 0, 0, 0, 0, 0, 0];

    // ---- Greeting -------------------------------------------------------------------------

    [Fact]
    public void GreetingOffersOnlyNoAuthWhenThereIsNoCredential()
    {
        var step = ForExample().Start();

        Assert.Equal(Socks5StepKind.Send, step.Kind);
        Assert.Equal([Socks5.Version, 1, (byte)Socks5Method.NoAuthentication], step.Bytes);
    }

    [Fact]
    public void GreetingOffersUsernamePasswordOnlyWhenACredentialExists()
    {
        var step = ForExample(new Socks5Credential("user", "pass")).Start();

        Assert.Equal(
            [Socks5.Version, 2, (byte)Socks5Method.NoAuthentication, (byte)Socks5Method.UsernamePassword],
            step.Bytes);
    }

    [Fact]
    public void StartTwiceIsAProtocolViolation()
    {
        var negotiator = ForExample();
        negotiator.Start();

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Start());
        Assert.Equal(Socks5ErrorCode.ProtocolViolation, error.Code);
    }

    [Fact]
    public void ReceiveBeforeStartIsAProtocolViolation()
    {
        var error = Assert.Throws<Socks5Exception>(() => ForExample().Receive([Socks5.Version, 0]));
        Assert.Equal(Socks5ErrorCode.ProtocolViolation, error.Code);
    }

    // ---- No-auth path ---------------------------------------------------------------------

    [Fact]
    public void NoAuthSelectionSendsTheConnectRequest()
    {
        var negotiator = ForExample();
        negotiator.Start();

        var step = negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        Assert.Equal(Socks5StepKind.Send, step.Kind);

        var expected = new List<byte>
        {
            Socks5.Version, (byte)Socks5Command.Connect, 0x00, (byte)Socks5AddressType.Domain, 11,
        };
        expected.AddRange(Encoding.UTF8.GetBytes("example.com"));
        expected.AddRange([0x01, 0xBB]); // 443

        Assert.Equal(expected, step.Bytes);
    }

    [Fact]
    public void SuccessfulConnectEstablishesTheTunnel()
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var step = negotiator.Receive(ConnectReply());

        Assert.Equal(Socks5StepKind.Established, step.Kind);
        Assert.Equal(Socks5Negotiator.State.Established, negotiator.CurrentState);
        Assert.Equal(Socks5Method.NoAuthentication, step.Connection!.Method);
        Assert.Empty(step.Connection.LeftoverBytes);
    }

    // ---- Partial delivery ------------------------------------------------------------------

    [Fact]
    public void AMethodSelectionSplitAcrossTwoReadsStillParses()
    {
        var negotiator = ForExample();
        negotiator.Start();

        Assert.Equal(Socks5StepKind.NeedMoreBytes, negotiator.Receive([Socks5.Version]).Kind);
        Assert.Equal(Socks5StepKind.Send, negotiator.Receive([(byte)Socks5Method.NoAuthentication]).Kind);
    }

    [Fact]
    public void AConnectReplyDeliveredOneByteAtATimeStillParses()
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var reply = ConnectReply();
        for (var i = 0; i < reply.Length - 1; i++)
        {
            Assert.Equal(Socks5StepKind.NeedMoreBytes, negotiator.Receive([reply[i]]).Kind);
        }

        Assert.Equal(Socks5StepKind.Established, negotiator.Receive([reply[^1]]).Kind);
    }

    [Fact]
    public void TruncationAtEveryOffsetIsRecoverable()
    {
        var reply = ConnectReply();

        for (var cut = 1; cut < reply.Length; cut++)
        {
            var negotiator = ForExample();
            negotiator.Start();
            negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

            Assert.Equal(Socks5StepKind.NeedMoreBytes, negotiator.Receive(reply.AsSpan(0, cut)).Kind);
            Assert.Equal(Socks5StepKind.Established, negotiator.Receive(reply.AsSpan(cut)).Kind);
        }
    }

    [Fact]
    public void BytesGluedToTheConnectReplyAreHandedBackNotDropped()
    {
        // A server may coalesce its reply with the first bytes of the relayed stream into one
        // segment. Dropping them looks like a rare, unreproducible truncation of the first response.
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var payload = "HTTP/1.1 200 OK"u8.ToArray();
        var step = negotiator.Receive([.. ConnectReply(), .. payload]);

        Assert.Equal(Socks5StepKind.Established, step.Kind);
        Assert.Equal(payload, step.Connection!.LeftoverBytes);
    }

    // ---- Authentication ---------------------------------------------------------------------

    [Fact]
    public void UsernamePasswordSubNegotiationUsesItsOwnVersionByte()
    {
        var negotiator = ForExample(new Socks5Credential("user", "pass"));
        negotiator.Start();

        var step = negotiator.Receive(MethodSelection(Socks5Method.UsernamePassword));

        // 0x01, not 0x05 — RFC 1929 versions itself independently.
        Assert.Equal(Socks5.AuthSubnegotiationVersion, step.Bytes![0]);
        Assert.Equal(
            [0x01, 4, (byte)'u', (byte)'s', (byte)'e', (byte)'r', 4, (byte)'p', (byte)'a', (byte)'s', (byte)'s'],
            step.Bytes);
    }

    [Fact]
    public void AuthenticationSuccessProceedsToConnect()
    {
        var negotiator = ForExample(new Socks5Credential("user", "pass"));
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.UsernamePassword));

        var step = negotiator.Receive([Socks5.AuthSubnegotiationVersion, 0x00]);

        Assert.Equal(Socks5StepKind.Send, step.Kind);
        Assert.Equal(Socks5Negotiator.State.AwaitingConnectReply, negotiator.CurrentState);
    }

    [Fact]
    public void AServerEchoingVersionFiveInTheAuthReplyIsTolerated()
    {
        // A known-common server bug. The status byte is what carries the decision.
        var negotiator = ForExample(new Socks5Credential("user", "pass"));
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.UsernamePassword));

        Assert.Equal(Socks5StepKind.Send, negotiator.Receive([Socks5.Version, 0x00]).Kind);
    }

    [Fact]
    public void AuthenticationFailureIsReportedAsSuch()
    {
        var negotiator = ForExample(new Socks5Credential("user", "wrong"));
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.UsernamePassword));

        var error = Assert.Throws<Socks5Exception>(() =>
            negotiator.Receive([Socks5.AuthSubnegotiationVersion, 0x01]));

        Assert.Equal(Socks5ErrorCode.AuthenticationFailed, error.Code);
        Assert.Equal(Socks5Negotiator.State.Failed, negotiator.CurrentState);
    }

    [Fact]
    public void SelectingUsernamePasswordWithNoCredentialIsRefused()
    {
        var negotiator = ForExample();
        negotiator.Start();

        var error = Assert.Throws<Socks5Exception>(() =>
            negotiator.Receive(MethodSelection(Socks5Method.UsernamePassword)));

        Assert.Equal(Socks5ErrorCode.AuthenticationRequired, error.Code);
    }

    [Fact]
    public void AnOverlongCredentialIsRejectedBeforeAnythingIsSent()
    {
        var negotiator = ForExample(new Socks5Credential(new string('u', 256), "pass"));

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Start());

        Assert.Equal(Socks5ErrorCode.CredentialTooLong, error.Code);
    }

    [Fact]
    public void CredentialRedactsItsPasswordInEveryTextualForm()
    {
        var credential = new Socks5Credential("user", "hunter2");

        Assert.DoesNotContain("hunter2", credential.ToString(), StringComparison.Ordinal);
        Assert.DoesNotContain("hunter2", $"{credential}", StringComparison.Ordinal);
    }

    // ---- Rejections --------------------------------------------------------------------------

    [Fact]
    public void NoAcceptableMethodsIsReportedAsSuch()
    {
        var negotiator = ForExample();
        negotiator.Start();

        var error = Assert.Throws<Socks5Exception>(() =>
            negotiator.Receive(MethodSelection(Socks5Method.NoAcceptableMethods)));

        Assert.Equal(Socks5ErrorCode.NoAcceptableAuthenticationMethod, error.Code);
    }

    [Fact]
    public void AMethodNeverOfferedIsRefused()
    {
        var negotiator = ForExample();
        negotiator.Start();

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Receive([Socks5.Version, 0x01]));

        Assert.Equal(Socks5ErrorCode.UnsupportedAuthenticationMethod, error.Code);
    }

    [Fact]
    public void AWrongVersionByteIsRefused()
    {
        var negotiator = ForExample();
        negotiator.Start();

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Receive([0x04, 0x00]));

        Assert.Equal(Socks5ErrorCode.UnexpectedVersion, error.Code);
    }

    [Theory]
    [InlineData(Socks5ReplyCode.GeneralFailure)]
    [InlineData(Socks5ReplyCode.ConnectionNotAllowed)]
    [InlineData(Socks5ReplyCode.NetworkUnreachable)]
    [InlineData(Socks5ReplyCode.HostUnreachable)]
    [InlineData(Socks5ReplyCode.ConnectionRefused)]
    [InlineData(Socks5ReplyCode.TtlExpired)]
    [InlineData(Socks5ReplyCode.CommandNotSupported)]
    [InlineData(Socks5ReplyCode.AddressTypeNotSupported)]
    public void EveryRejectionCodeIsSurfacedVerbatim(Socks5ReplyCode code)
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Receive(ConnectReply(code)));

        Assert.Equal(Socks5ErrorCode.RequestRejected, error.Code);
        Assert.Equal(code, error.ReplyCode);
    }

    [Fact]
    public void AnUnknownReplyCodeIsNamedRatherThanGuessedAt()
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var error = Assert.Throws<Socks5Exception>(() =>
            negotiator.Receive([Socks5.Version, 0x42, 0x00, 0x01, 0, 0, 0, 0, 0, 0]));

        Assert.Equal(Socks5ErrorCode.UnknownReplyCode, error.Code);
    }

    [Fact]
    public void ANonZeroReservedByteIsMalformed()
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var error = Assert.Throws<Socks5Exception>(() =>
            negotiator.Receive([Socks5.Version, 0x00, 0x99, 0x01, 0, 0, 0, 0, 0, 0]));

        Assert.Equal(Socks5ErrorCode.MalformedResponse, error.Code);
    }

    [Fact]
    public void AnUnknownAddressTypeInTheReplyIsRefused()
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var error = Assert.Throws<Socks5Exception>(() =>
            negotiator.Receive([Socks5.Version, 0x00, 0x00, 0x07, 0, 0, 0, 0, 0, 0]));

        Assert.Equal(Socks5ErrorCode.UnsupportedAddressType, error.Code);
    }

    [Fact]
    public void AHostileDomainLengthCanOnlyStall()
    {
        // The server chooses the length byte. The bounds check means an absurd value produces
        // "waiting for more bytes", never an over-read.
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));

        var step = negotiator.Receive([Socks5.Version, 0x00, 0x00, (byte)Socks5AddressType.Domain, 0xFF, 0x41]);

        Assert.Equal(Socks5StepKind.NeedMoreBytes, step.Kind);
    }

    [Fact]
    public void ReceiveAfterFailureIsRefused()
    {
        var negotiator = ForExample();
        negotiator.Start();
        Assert.Throws<Socks5Exception>(() => negotiator.Receive([0x04, 0x00]));

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Receive([0x05, 0x00]));
        Assert.Equal(Socks5ErrorCode.ProtocolViolation, error.Code);
    }

    [Fact]
    public void ReceiveAfterEstablishedIsRefused()
    {
        var negotiator = ForExample();
        negotiator.Start();
        negotiator.Receive(MethodSelection(Socks5Method.NoAuthentication));
        negotiator.Receive(ConnectReply());

        var error = Assert.Throws<Socks5Exception>(() => negotiator.Receive([0x00]));
        Assert.Equal(Socks5ErrorCode.ProtocolViolation, error.Code);
    }

    // ---- Transient classification -------------------------------------------------------------

    [Theory]
    [InlineData(Socks5ReplyCode.HostUnreachable, true)]
    [InlineData(Socks5ReplyCode.NetworkUnreachable, true)]
    [InlineData(Socks5ReplyCode.TtlExpired, true)]
    [InlineData(Socks5ReplyCode.ConnectionNotAllowed, false)]
    [InlineData(Socks5ReplyCode.CommandNotSupported, false)]
    public void RejectionTransienceIsClassified(Socks5ReplyCode code, bool transient) =>
        Assert.Equal(transient, Socks5Exception.RequestRejected(code).IsTransient);

    [Fact]
    public void AuthenticationFailureIsNotTransient() =>
        Assert.False(Socks5Exception.AuthenticationFailed().IsTransient);
}
